### PR TITLE
Move Location up to SyntaxNode.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-compiler 

As part of my SyntaxRefactoring (https://github.com/dotnet/roslyn/pull/13160) i ran into something interesting that prompted this PR.

1. the change i'm working on pulls all the syntax-list types into Core so that we don't have duplicated versions of them for C# and for VB.  
2. Because the nodes are in Core they cannot subclass CSharpSyntaxNode or VBSyntaxNode.  Instead, the lists subclass SyntaxNode itself.
3. In the Bound-Node system we have BoundNodes point at SyntaxNodes.
4. There are Bound-Nodes that correspond to syntax-lists (for example, BoundStatementList)
5. Because of this, it was no longer possible for BoundNode to point at CSharpSyntaxNode.  Instead, it needed to be weakened to point at a SyntaxNode instead.

Doing this i discovered that the parts of the system that used BoundNodes (and their SyntaxNodes) never needed the nodes to be CSharpSyntaxNodes.  So i've relaxed this whole part of the system.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler If you can review this quickly, i would appreciate it.  It's a change to a lot of files, and the longer it stays open, the more chance of merge conflicts i'll have to deal with.  Thanks!

**Comment by CyrusNajmabadi** [comment]
Note: i have a continuation of this PR that also makes these changes for VB here: https://github.com/dotnet/roslyn/pull/13168

File: src/Compilers/CSharp/Portable/Binder/LocalBinderFactory.cs [hunk 1]
```diff
@@ -49,15 +49,16 @@ private void Visit(CSharpSyntaxNode syntax, Binder enclosing)
...
             var builder = new LocalBinderFactory(containingMemberOrLambda, syntax, enclosing, methodsWithYields);
 
-            if (syntax is ExpressionSyntax)
+            var exprNode = syntax as ExpressionSyntax;
+            if (exprNode != null)
             {
                 var binder = new ExpressionVariableBinder(syntax, enclosing);
                 builder.AddToMap(syntax, binder);
-                builder.Visit(syntax, binder);
+                builder.Visit(exprNode, binder);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/LocalBinderFactory.cs]: Was this change intentional?

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Binder/LocalBinderFactory.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Binder/LocalBinderFactory.cs]: Yes.  'Visit' Can only be called on a CSaarpSyntaxNode (the 'Visit/Accept' methods are not on SyntaxNode itself).  However, as we can see from this section, we only ever called Visit on a node that was known to be an ExpressionSyntax.  So, in this code, i would have had to cast (safely) the syntax node to a CSharpSyntaxNode. Instead of having an is+cast, i just replaced both with a single 'as' cast.

**Comment by jcouv** [comment]
LGTM

**Comment by CyrusNajmabadi** [comment]
Closing in favor of: https://github.com/dotnet/roslyn/pull/13168
